accord-core/src/main/java/accord/topology/TopologyManager.java

```diff
@@ -371,17 +395,86 @@ private int indexOf(long epoch)
         }
     }
 
+    private static class FutureEpoch
+    {
+        private volatile AsyncResult.Settable<Void> future;
+        private long deadlineMillis;
+
+        public FutureEpoch(long deadlineMillis)
+        {
+            this.future = AsyncResults.settable();
+            this.deadlineMillis = deadlineMillis;
+        }
+
+        /*
+         * Notify any listeners that are waiting for the epoch that is has been a long time since
+         * we started waiting for the epoch. We may still eventually get the epoch so also create
+         * a new future so subsequent operations may have a chance at seeing the epoch if it ever appears.
+         *
+         * Subsequent waiters may get a timeout notification far sooner (WATCHDOG_INTERVAL_MILLISS)
+         * instead of EPOCH_INITIAL_TIMEOUT_MILLIS
+         */
+        @GuardedBy("TopologyManager.this")
+        public void timeOutCurrentListeners(long newDeadline, Agent agent)
+        {
+            deadlineMillis = newDeadline;
+            AsyncResult.Settable<Void> oldFuture = future;
+            if (oldFuture.isDone())
+                return;
+            future = AsyncResults.settable();
+            future.addCallback(agent);
```

Comment:
this is unsafe (check-then-set pattern), if the `tryFailure` returns `false` it means the future got resolved *after* the `isDone` check, and in this case we would leak and case this new future to produce a timeout even though the epoch is ready.

Reply:
Is it actually a race? They are guarded by the same lock so they aren't racing.

Reply:
looks like its set in `accord.topology.TopologyManager#onTopologyUpdate`, so both share the same lock; so this code is fine